### ISSUE TITLE
Emergency issue - halfway through a campaign, all emails suddenly going to me.

### ISSUE THREAD

**Issue by bossanova808** [putyourlightson/craft-campaign#350]
Craft 3, latest, plus latest Campaign for Craft 3.

Our newsletter has just gone completely haywire. 43% through sending (about 4100 customers out of 9000), **I** suddenly started receiving hundreds of copies of the newsletter.

All personalised with the customer names in them, but all sent to my email address.

Looking at Postmark activity - they are being sent to me, not to the corresponding customer.  That's good, at least, as I was panicking that customers were getting hundreds of copies...

I paused the campaign - but still they keep arriving.
I cancelled the campaign -  but still they keep arriving.

I currently have 450+ copies and they just won't stop - I even rebooted the bloody server.

There's no job running in Craft queues to be seen.  The sendout _says_ it is cancelled.

But it just won't die!!  Help!!  The email just keep coming...

**Comment by bossanova808** [comment]
Have just removed the Postmark API key from Campaign's settings.

Emails are STILL coming through - which must mean it's something at Postmark - it's processing a queue of messages at its end and that is still going??

**Comment by bossanova808** [comment]
(just to be clear, no one was doing anything on the server or poking around settings or anything at the time)

**Comment by bossanova808** [comment]
Still going, at over 750 copies to me now.  Have contacted Postmark support.

I have confirmed removing the API key 'worked' - the test email fails.  I can't see any queue activity at all, and there's nothing in the Campaign logs except:

2022-12-07 17:19:46 [][1][info] Sendout "2022_12_07 - December 2022 Newsletter" paused by "xxx@imagescience.com.au".
2022-12-07 17:23:55 [][1][info] Sendout "2022_12_07 - December 2022 Newsletter" cancelled by "xxx@imagescience.com.au".

**Comment by bossanova808** [comment]
Postmark added my email to their suppression list, which stopped things for now.    They're confident there can't be anything at their end causing this.

After a while, they re-enabled my email, as the queue has cleared.  Emails have not, thank God, resumed.  I ended up with about 800 copies.

They reported that the emails in the raw source had this:

![image](https://user-images.githubusercontent.com/731309/206114742-64991c08-c85e-401a-94c2-8983054e106b.png)

And asked if 'Test Recipient' was a clue...which I suppose it probably is.

But what can possibly have happened, in the middle of a Campaign, with no-one doing anything related on the server - to cause this???

**Comment by bossanova808** [comment]
And how can I revert all this, such that I can re-send that campaign, only to folks who didn't get it, and also be sure this won't happen again?

(I apologise for the mid emergency stream of consciousn3ss emails - but I was driving home when my notification suddenly blew up with hundreds of emails coming in...)

**Comment by bencroker** [comment]
If you're using any queue job runners that are out of the ordinary, ensure that those processes are killed. Otherwise, this sounds very unusual, especially if you rebooted the server and removed the API keys...

**Comment by bencroker** [comment]
Sorry to hear about your issue and the panic it caused!

It sounds like the [testToEmailAddress](https://craftcms.com/docs/4.x/config/general.html#testtoemailaddress) config setting may have been set in the middle of the sendout sending, is that possible?

Otherwise, do you have any queue job handlers set up that are not the Craft default? It may be possible that a rogue PHP process got carried away.

Also, what specific version of Campaign are you running and did you update recently or since the last sendout?

**Comment by bossanova808** [comment]
Craft Pro 3.7.61
Campaign 1.24.0

```
commit fd7f3ecfe9fee5318f13c445d8b85261b3fbc7ab
Date:   Mon Nov 21 15:33:16 2022 +1100
    [UP] craft-seomatic (3.4.42 => 3.4.43) craft-campaign (1.23.6 => 1.24.0)
```

Last campaign to this list was 22nd of November - i.e. the day _after_ I last updated Campaign (I update everything to the latest available pretty regularly, albeit Craft 3 latest available...).  No issues with that one.

Queue running - I have a cron job calling the sendouts thing:

```
# Cronjob for Campaign plugin - moving off the hour here as queue errors...
# https://github.com/putyourlightson/craft-campaign/issues/143
5,17,25,35,49,55 * * * * cronic php /srv/users/serverpilot/apps/imagescience/craft/craft campaign/sendouts/run
```

And my queue running is done via a systemd approach - https://nystudio107.com/blog/robust-queue-job-handling-in-craft-cms#solution-4-utilizing-systemd

I have this set in my general.php:

```
        // https://docs.craftcms.com/v3/config/config-settings.html#testtoemailaddress
        'testToEmailAddress' => (\craft\helpers\App::env('SEND_ALL_EMAILS_TO')) ?? null,
```

No one has access to this server (at the config level, there are people processing orders or writing entries) other than me - and I deploy using Ansible (and the machine I do that from was switched off, with me in my car when this happened..) - so there is no possibility of some lingering cloud deploy thing having access to the server.

On my production server, `SEND_ALL_EMAILS_TO` is **definitely** NOT set in the .env, and I have carefully grepped the entire server just to be sure.  It simply doesn't exist in the .env file on production,  so there's no way it can have been set accidentally to on.  It _is_ set on my dev server - but the .env stuff is not part of any deploys etc (and there's zero chance of any deploys being run when this happened as...see above...)

I can also be sure that that was not set, as all _other_ Craft emails (for online orders etc) were being sent correctly to customers and not me, during the same period this was happening (confirmed by checking the Postmark logs).  Note those emails use the Transactional stream, and _only_ Campaign uses the Broadcast stream.   It's definitely only Campaign emails that went crazy, and definitely at the same time as other emails were being sent by Craft correctly.

Looking at the postmark logs, the trouble starts at 5:15:06pm - I can see the first email to me, intermixed with emails to customers.  For the next 5 mins or so, I can see both customers and me getting copies, and then from about 5:20pm on it's 100% me for the next 40 mins or so until Postmark suppressed my address and the queue cleared (along the way I disabled that cronjob too, now re-enabled).

I've just got up over here and will do some digging this morning, but sending all this now for your info.

**Comment by bossanova808** [comment]
Have carefully checked the `syslog` and there are no oddities in there anywhere near the times in question - I can see the various batches starting and all looks completely normal 

```
Dec  7 17:13:04 production nice[3909]: 2022-12-07 17:13:04 [6644] Sending “2022_12_07 - December 2022 Newsletter” sendout [batch 71] (attempt: 1, pid: 3909) - Done (144.307 s, 69.32 MiB)
Dec  7 17:13:16 production nice[3908]: 2022-12-07 17:13:16 [6645] Sending “2022_12_07 - December 2022 Newsletter” sendout [batch 72] (attempt: 1, pid: 3908) - Started
```

...and there's no other stuff (ssh logins, or ...well, anything really).

I can see the reboot of the server, and then I see no more notifications of the sendout batches after that - showing that after rebooting the server (at the latest) - things must have been in the queue at Postmark at that point, and carried on for a while after (about half an hour, which I find surprising!).

This leads me to think I need to look a bit earlier than when the emails started coming to me, as there's clearly a lag at the postmark end of things.

**Comment by bossanova808** [comment]
For the life of me I have no clue whatsoever as to what happened here.  I can't find anything that looks suspicious or related, anywhere.  

Even if, and it's a truly zero chance if (as far as I can determine), there was e.g. a brief problem with Craft accessing the production .env, I can't see how it would fall back to having `testToEmailAddress` set.  Especially while Craft itself is simultaneously sending emails to the Transaction stream without issue.  

I have asked Postmark to confirm that all the 800 odd emails had 'Test Recipient' set as the name.  My batch size in Campaign is set to 1000, although it will be the 180 second PHP limit that effectively limits the batch size (I don't use caching on my campaign templates so it's slow - we've discussed this before, I just haven't had a chance to go back and deal with that properly, and don't really mind if it takes a few hours to send this out, really).

I do know that before this a few thousand emails DID go out ok (as I can see the delivery and open/click reports for all of those in Postmark).  So there is _some_ brief mystery event when one/a few of these batches was created.

I think I want to try and trigger the completion of this sendout.  If it sends the rest out correcrly, I can live with those 800 people not getting this one will put this down to Gremlins for now as I just don't have much time.  

I don't think that is possible with a Cancelled one, via the UI in Campaign, right?  Can I manually change the sendout status in the db perhaps, to get it to restart the send, to those not yet sent to?  If so - what is the right status value for that?

I will watch it like a hawk when I do this, of course...

**Comment by bossanova808** [comment]
I am thinking change the status to 'paused' in the db and then I can manually kick it off again from the GUI - is that too crazy an idea?

I see the statusses seem to be:

```
    const STATUS_SENT = 'sent';
    const STATUS_SENDING = 'sending';
    const STATUS_QUEUED = 'queued';
    const STATUS_PENDING = 'pending';
    const STATUS_PAUSED = 'paused';
    const STATUS_CANCELLED = 'cancelled';
    const STATUS_FAILED = 'failed';
    const STATUS_DRAFT = 'draft';
```

**Comment by bencroker** [comment]
> I am thinking change the status to 'paused' in the db and then I can manually kick it off again from the GUI - is that too crazy an idea?

That should work just fine. It might be smart to send a campaign to a small test list first, ensuring that emails are sent to the intended recipients, and not you.

**Comment by bencroker** [comment]
Does "Test Recipient" exist as a contact? And is it by any chance set as one of the "Default Test Contacts" in the campaign type's settings?

**Comment by bossanova808** [comment]
That's a nope:

![image](https://user-images.githubusercontent.com/731309/206335049-62b1062a-3582-4b5f-b623-961481dc6aa8.png)

And I don't have anything set for that setting for that Campaign:

![image](https://user-images.githubusercontent.com/731309/206335152-d3e236e3-310d-4c79-a963-1dab380a3817.png)

I've just done a test run to our internal list here, and that worked fine.  I'm obviously a bit nervous about kicking the main one off again, given I have no explanation for this, but...not much else I can do really, is there??

**Comment by bencroker** [comment]
Well, we're still working through it here. So you have no idea where "Test Recipient" came from? What about the email address that received all those emails, does it exist as a contact in the system?

**Comment by bossanova808** [comment]
No - I grepped the entire Project for `Test Recipient` and only Craft config stuff came up (i.e. not me setting anything, just the comments etc in the vendor folder).

Yes that email address is mine, it's my primary email, my Craft login user email etc.  And it definitely has my name and not 'Test Recipient'

Honestly, I didn't even think Campaign really used Craft's email stuff at all - I though it talked to Postmark via the API?  In which case - it seems bizarre that it would have that name.  But I definitely don't have a contact with that name...

Note Postmark have not got back to me yet to confirm _all_ the emails has that 'Test Recipient' (I've just seen that one screenshot) - but it was definitely a newsletter send (based on the subject shown in the screenshot) and not any actual test email, so I think reasonable to assume they all had that name...

**Comment by bencroker** [comment]
> Yes that email address is mine, it's my primary email, my Craft login user email etc. And it definitely has my name and not 'Test Recipient'

You didn't answer my question though: does it exist as a contact in the system?

**Comment by bossanova808** [comment]
Sorry; yes it's definitely in the system as a Campaign Contact.

**Comment by bencroker** [comment]
Tracked it down to this method. I'm now convinced that `testToEmailAddress` was somehow set:
https://github.com/craftcms/cms/blob/615c267a496dcacb6e008130590dfc4b690f61ec/src/config/GeneralConfig.php#L2071-L2084

**Comment by bossanova808** [comment]
Well,  yep,that seems the likeliest/only opton -  I use it on DEV to test order emails etc - it's permanently set on dev.  

But I just don't see how it can spontaneously enable on a live server when no one is anywhere near any config etc.

Anyhow, I am going to guess there was something errant lingering somewhere, I just can't even imagine what.  Ansible all runs and terminates on the command line and even _if_ there was somehow a brief moment where the .env was somehow still being updated from a lingerging previous deploy thing - then Craft itself should pretty much go down, as it couldn't get to the DB etc.

Honestly bizarre.

I'm going to proceed with kicking it off again and pray to the Gods...

**Comment by bencroker** [comment]
Could the emails possibly have originated from your dev server (or a staging server) and not the production server, after all? That seems much more plausible than something spontaneously changing in the production environment.

**Comment by bossanova808** [comment]
Hmm, ok, that is perhaps an interesting question.  And I agree, much more plausible.

So - I can't _recall_ doing this yesterday, but it's not impossible I did (I'll have to dig through some logs) - but IF I pulled a copy of the production DB down to dev while the sending was occurring, I suppose ... would it then run on the dev server (it also has a systemd queue runner etc, but no cron job to kick things off)??  I certainly did not have this on my radar as a possibility.  I guess I don't really understand exactly what kicks off the send then, as on the dev server nothing will have called campaign/sendouts/run - but I guess it syncs the queues too, as ultimately they are in the db too?  In which case - I should not be syncing that table I suppose (?!).

This is both an explanation,  I guess - and good news, as it means the production server I can jsut restart, and then all people will still get their copies...

**Comment by bencroker** [comment]
Check if a sendout exists with recipients in the `dev` environment. You can also drill into the `campaign_contacts_campaigns` table to see exact timings, if it helps.

**Comment by bossanova808** [comment]
Yep, so looking at my bash history, which was the easiest in the end - I can see I did a deploy (this was before sending the newsletter) - and I then not long after pulled the DB as I made a very small field change (updated some field instructions) - and thanks to the horror show that is Project Config, I pretty much automatically pull a DB copy after that, as otherwise I will lose those changes on the next deploy.  

I must have fired off the newsletter before that had  got to the bit where it dumps the production DB - the process takes a while - as this is a big production DB by now, and I don't really watch it.  

So I will definitely have pulled in an entry in the queue table then.  And that has resulted in this.  Mystery solved, although I am still quite surprised that Craft carried on with this on the dev server, but I suppose it's just a record in a table.  

Very confusing.  And I have now made sure I _won't_ be syncing that table in the future (well, the structure, just nothing in it...).

I can confirm the sendout has been resumed and is proceeding on production, all according to plan, at this moment.
